applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp

```diff
@@ -157,14 +138,35 @@ class DivDGMaxEigenvalue<DIM>::SolverWorkspace {
      */
     void writeEigenvectorAsTimeIntegrationVector(
         std::size_t eigenvalue, std::size_t timeIntegrationVector) {
-        logger.assert_debug(eigenvalue < numberOfConvergedEigenpairs,
+        logger.assert_debug(eigenvalue < eigenpairs_.size(),
                             "Eigenvalue index too large");
         PetscErrorCode err;
-        err = VecCopy(eigenvectors_[eigenvalue], tempVector_);
+        err = VecCopy(eigenpairs_.getEigenvector(eigenvalue), tempVector_);
         CHKERRABORT(PETSC_COMM_WORLD, err);
         tempVector_.writeTimeIntegrationVector(timeIntegrationVector);
     }
 
+    LinearAlgebra::MiddleSizeMatrix computeOverlapIntegrals() {
+        LinearAlgebra::MiddleSizeMatrix result(eigenpairs_.size(),
+                                               previousEigenpairs_.size());
+        PetscErrorCode err;
+        for (std::size_t i = 0; i < eigenpairs_.size(); ++i) {
+            err = MatMult(massMatrix_, eigenpairs_.getEigenvector(i),
+                          tempVector_);
```

Comment:
Is the member tempVector_ necessary for performance? otherwise this method could be `const`

Reply:
It is needed because `MatMult(M, x, y)`  corresponds to `y = Mx` has the requirement that x and y must not be the same vector. Which I guess is indeed for performance reasons.

What I could do is make `tempVector_` mutable, which sounds quite reasonable as it is just a vector for temporary storage.

Reply:
Looking further into it. Unfortunately all the PETSc and const don't really work well together. All the types (Mat, Vec, etc) are typdef's to pointer types (i.e. `typedef Vec somestruct*`). So making them all const (e.g. `const Vec`) would only make the pointer const, not the target matrix, vector, etc. So I don't see a practical way to make it more const safe.